### PR TITLE
Added btcmarkets streaming code

### PR DESCRIPTION
**Pull request by nielsdraaisma**

### CODE REVIEW

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingMarketDataService.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    return BTCMarketsStreamingAdapters.adaptOrderbookMessageToOrderbook(message);
+  }
+
+  @Override
+  public Observable<OrderBook> getOrderBook(CurrencyPair currencyPair, Object... args) {
+    String marketId = BTCMarketsStreamingAdapters.adaptCurrencyPairToMarketId(currencyPair);
+    return service
+        .subscribeChannel(CHANNEL_ORDERBOOK, marketId)
+        .map(node -> mapper.treeToValue(node, BTCMarketsWebSocketOrderbookMessage.class))
+        .filter(orderEvent -> orderEvent.marketId != null && orderEvent.marketId.equals(marketId))
```
Comment (badgerwithagun) [inline xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingMarketDataService.java:39]: could simplify this to `marketId.equals(orderEvent.marketId)` and lose the null check

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import info.bitrich.xchangestream.btcmarkets.dto.BTCMarketsWebSocketSubscribeMessage;
+import info.bitrich.xchangestream.service.netty.JsonNettyStreamingService;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BTCMarketsStreamingService extends JsonNettyStreamingService {
+  public static final String CHANNEL_ORDERBOOK = "orderbook";
```
Comment (badgerwithagun) [inline xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java:15]: Do these need to be `public`?

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import com.google.common.collect.Sets;
+import info.bitrich.xchangestream.btcmarkets.dto.BTCMarketsWebSocketSubscribeMessage;
+import info.bitrich.xchangestream.service.netty.JsonNettyStreamingService;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BTCMarketsStreamingService extends JsonNettyStreamingService {
```
Comment (badgerwithagun) [inline xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java:14]: Does this class itself need to be `public`?

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    if (CHANNEL_ORDERBOOK.equals(channelName)) {
+      return channelName + ":" + args[0].toString();
+    }
+    return channelName;
+  }
+
+  @Override
+  public String getUnsubscribeMessage(String channelName) throws IOException {
+    if (channelName.startsWith(CHANNEL_ORDERBOOK)) {
+      final String[] parts = channelName.split(":");
```
Comment (badgerwithagun) [inline xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java:64]: Looks like there should be a `subscribedOrderbooks.remove()` here.

nit: Setting up `subscribedOrderbooks` in the `get...Message` methods seems unsafe, but I can't see an obvious cleaner way. It's not the ideal websocket API, so I'm not too bothered.

File: xchange-stream-btcmarkets/src/test/resources/log4j2.xml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (nielsdraaisma) [reply xchange-stream-btcmarkets/src/test/resources/log4j2.xml]: Removed that.

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/dto/BTCMarketsWebSocketOrderbookMessage.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+  public final List<List<String>> bids;
+  public final List<List<String>> asks;
+
+  public final String messageType;
+
+  public BTCMarketsWebSocketOrderbookMessage(
+      @JsonProperty("marketId") String marketId,
+      @JsonProperty("timestamp") String timestamp,
+      @JsonProperty("bids") List<List<String>> bids,
```
[the remark this answers is not in the document]
Comment (nielsdraaisma) [reply xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/dto/BTCMarketsWebSocketOrderbookMessage.java]: Done

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/BTCMarketsStreamingAdapters.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.Order;
+import org.knowm.xchange.dto.marketdata.OrderBook;
+import org.knowm.xchange.dto.trade.LimitOrder;
+import org.knowm.xchange.utils.DateUtils;
+
+public class BTCMarketsStreamingAdapters {
```
[the remark this answers is not in the document]
Comment (nielsdraaisma) [reply xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/BTCMarketsStreamingAdapters.java]: Done

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingMarketDataService.java (hunk 1 above)
Comment (nielsdraaisma) [reply xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingMarketDataService.java]: Done

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java (hunk 4 above)
Comment (nielsdraaisma) [reply xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java]: Done, there's no clean way of working around that, an unsubscribe is basically subscribing again without that channel.

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingDigest.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Base64;
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import org.knowm.xchange.utils.Assert;
+
+public class BTCMarketsStreamingDigest {
```
[the remark this answers is not in the document]
Comment (nielsdraaisma) [reply xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingDigest.java]: I've removed that class entirely, i'll raise a new PR later with working private channels, that's not currently used at all.

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java (hunk 2 above)
Comment (nielsdraaisma) [reply xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java]: I've moved things out of the packages and removed  `public`.

File: xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java (hunk 3 above)
Comment (nielsdraaisma) [reply xchange-stream-btcmarkets/src/main/java/info/bitrich/xchangestream/btcmarkets/service/BTCMarketsStreamingService.java]: I've moved things out of the packages and removed  `public`.

**Comment by nielsdraaisma** [comment]
Thanks @badgerwithagun for your review, aside from some visibility issues i've addressed the issues you raised. I've moved things around a bit and removed the public visibility of some classes.

**Pull request merged**